src/health_checker.cc

```diff
@@ -39,12 +39,31 @@ void HealthChecker::SetUnhealthy(const std::string& component) {
 
 bool HealthChecker::IsHealthy() const {
   std::lock_guard<std::mutex> lock(mutex_);
-  return unhealthy_components_.empty();
+  if (!unhealthy_components_.empty()) {
+    return false;
+  }
+  for (auto& c : component_callbacks_) {
```

Comment:
Should you not lock `mutex_` while iterating? It's possible to split this into read and write locks (https://en.cppreference.com/w/cpp/thread/shared_mutex).

Reply:
Ah, are you concerned about a possible deadlock? What's the right way to do this?